### PR TITLE
Add Cypress page for using `@clerk/testing`

### PR DESCRIPTION
**Pull request by anagstef**
Adds a page that explains how to use the `@clerk/testing` with Cypress.

> https://clerk.com/docs/pr/1053/testing/cypress

### CODE REVIEW

**Comment by alexisintech** [comment]
Hi! Tag me in once the appropriate people (I'm assuming @agis and @kylemac ) have reviewed and made sure the content is correct!
Once they're done, I'll review the writing 😊

File: docs/testing/cypress.mdx [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+Cypress is a well-established JavaScript testing framework. This guide aims to help you set up your environment for creating authenticated tests with Clerk. This guide will assume you're somewhat familiar with Clerk and Cypress.
+
+Before diving in, you might want to start by visiting the [Cypress starter](https://github.com/clerk/example-cypress-nextjs) for an example repo that tests a Clerk-powered application using [Testing Tokens](/docs/testing/overview#testing-tokens).
+
+<Steps>
+### Install `@clerk/testing`
+
+Clerk's testing package provides integration helpers for popular testing frameworks.
+
+```bash filename="terminal"
```
Comment (octoper) [inline docs/testing/cypress.mdx:21]: Is there any issue with including `yarn` and `pnpm` package managers as we do on other pages?

File: docs/testing/cypress.mdx [hunk 2]
```diff
@@ -0,0 +1,79 @@
+---
+title: Testing with Cypress
+description: Use Cypress to write end-to-end tests with Clerk.
+---
+
+# Testing with Cypress
+
+<Callout type="info">
```
Comment (octoper) [inline docs/testing/cypress.mdx:8]:
```suggestion
<Callout type="warning">
```
❓ Should we change that to a warning to be a bit more obvious?

**Review by octoper — APPROVED** [review]
💯

File: docs/testing/cypress.mdx (hunk 2 above)
Comment (alexisintech) [reply docs/testing/cypress.mdx:8]: I agree and please use our new syntax `> [!WARNING]`

File: docs/testing/cypress.mdx (hunk 1 above)
Comment (alexisintech) [reply docs/testing/cypress.mdx:21]: I agree with this - please see the Nextjs quickstart for reference. We use `CodeBlockTabs` to provide the different installation options: npm, yarn, pnpm

File: docs/testing/cypress.mdx [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+</Callout>
+
+Cypress is a well-established JavaScript testing framework. This guide aims to help you set up your environment for creating authenticated tests with Clerk. This guide will assume you're somewhat familiar with Clerk and Cypress.
+
+Before diving in, you might want to start by visiting the [Cypress starter](https://github.com/clerk/example-cypress-nextjs) for an example repo that tests a Clerk-powered application using [Testing Tokens](/docs/testing/overview#testing-tokens).
+
+<Steps>
+### Install `@clerk/testing`
+
+Clerk's testing package provides integration helpers for popular testing frameworks.
```
Comment (alexisintech) [inline docs/testing/cypress.mdx:19]:
```suggestion
Clerk's testing package provides integration helpers for popular testing frameworks. Install it by running the following command:
```

File: docs/testing/cypress.mdx [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+```bash filename="terminal"
+npm install @clerk/testing --save-dev
+```
+
+### Set your API keys
+
+Set your publishable and secret keys in your test runner, as the `CLERK_PUBLISHABLE_KEY` and `CLERK_SECRET_KEY` environment variables respectively.
```
Comment (alexisintech) [inline docs/testing/cypress.mdx:27]:
```suggestion
In your test runner, set your publishable and secret key as the `CLERK_PUBLISHABLE_KEY` and `CLERK_SECRET_KEY` environment variables, respectively.

To find your keys:
1. Navigate to the [Clerk Dashboard](https://dashboard.clerk.com/last-active?path=api-keys).
2. In the navigation sidebar, select **API Keys**.
3. In the **Quick Copy** section, copy your Clerk publishable and secret key.
```

File: docs/testing/cypress.mdx [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+<Callout type="warning">
+  Ensure you provide the secret key in a secure manner, to avoid leaking it to third parties. For example, if you are using GitHub Actions, refer to [*Using secrets in GitHub Actions*](https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions).
+</Callout>
```
Comment (alexisintech) [inline docs/testing/cypress.mdx:31]: Let's use our new syntax as to avoid tech debt!

```suggestion
> [!WARNING]
> Ensure you provide the secret key in a secure manner, to avoid leaking it to third parties. For example, if you are using GitHub Actions, refer to [*Using secrets in GitHub Actions*](https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions).
```

File: docs/testing/cypress.mdx [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+### Global setup
+
+In your [cypress.config.ts](https://docs.cypress.io/guides/references/configuration), call the `clerkSetup` function.
+
+<Callout type="info">
+    Keep in mind that you must pass the Cypress `config` object to the `clerkSetup` function and also return the new config object from the `setupNodeEvents` function.
+</Callout>
```
Comment (alexisintech) [inline docs/testing/cypress.mdx:39]: We don't have to explicitly explain exactly what the code is doing - users can see this for themselves in the example.

```suggestion
To set up Clerk with Cypress, add the following code to your [`cypress.config.ts`](https://docs.cypress.io/guides/references/configuration) file.
```

File: docs/testing/cypress.mdx [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+  }
+});
+```
+
+`clerkSetup` will obtain a [Testing Token](/docs/testing/overview#testing-tokens) when your test suite starts, making it available for all subsequent tests to use.
+
+<Callout type="info">
+  You can set the Testing Token yourself as opposed to calling `clerkSetup`, by
+  setting it in the `CYPRESS_CLERK_TESTING_TOKEN` environment variable.
+</Callout>
```
Comment (alexisintech) [inline docs/testing/cypress.mdx:60]: Does this mean they can skip this entire step if they provide this env var? If so, I can fix up this section

File: docs/testing/cypress.mdx [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+# Testing with Cypress
+
+<Callout type="info">
+  Our `@clerk/testing` package supports only end-to-end testing with Cypress. Unit tests are not supported.
+</Callout>
+
+Cypress is a well-established JavaScript testing framework. This guide aims to help you set up your environment for creating authenticated tests with Clerk. This guide will assume you're somewhat familiar with Clerk and Cypress.
+
+Before diving in, you might want to start by visiting the [Cypress starter](https://github.com/clerk/example-cypress-nextjs) for an example repo that tests a Clerk-powered application using [Testing Tokens](/docs/testing/overview#testing-tokens).
```
Comment (alexisintech) [inline docs/testing/cypress.mdx:14]: Let's explain the purpose of this guide better - we want to show users how to incorporate Clerk's testing tokens for their Cypress test. Testing tokens are the star of the show here, so we want to explain them and emphasize their usage.

```suggestion
Testing with Cypress requires the use of Clerk's [Testing Tokens](/docs/testing/overview#testing-tokens) to bypass various bot detection mechanisms that typically safeguard Clerk applications against malicious bots. Without Testing Tokens, you may encounter "Bot traffic detected" errors in your requests.

This guide aims to help you set up your environment for creating Clerk-authenticated tests with Cypress.

> [!NOTE] 
> Check out the [demo repo](https://github.com/clerk/example-cypress-nextjs) that tests a Clerk-powered application using [Testing Tokens](/docs/testing/overview#testing-tokens).
```

File: docs/testing/cypress.mdx [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+`clerkSetup` will obtain a [Testing Token](/docs/testing/overview#testing-tokens) when your test suite starts, making it available for all subsequent tests to use.
+
+<Callout type="info">
+  You can set the Testing Token yourself as opposed to calling `clerkSetup`, by
+  setting it in the `CYPRESS_CLERK_TESTING_TOKEN` environment variable.
+</Callout>
+
+### Use `setupClerkTestingToken`
+
+Then, in your tests use the `setupClerkTestingToken` function to augment your requests with the token:
```
Comment (alexisintech) [inline docs/testing/cypress.mdx:64]:
```suggestion
### Use Clerk Testing Tokens

Now that Cypress is configured with Clerk, you can use the `setupClerkTestingToken()` function in your requests to augment them with the Testing Token. See the following example:
```

File: docs/testing/cypress.mdx (hunk 7 above)
Comment (anagstef) [reply docs/testing/cypress.mdx:76]: Yes, you can skip this step if you provide a token. Although, I think this will confuse people because using the `clerkSetup` is the best/quickest way to get a token. So, I'll go ahead and completely remove this section.

**Comment by anagstef** [comment]
@alexisintech @octoper Thank you both for reviewing and adding suggestions for this PR!! 🙏  I applied all of your suggestions.

Also, I added a Custom Commands section, If you also want to review it! Thank you!
This PR will be ready to be merged, after this SDK release: https://github.com/clerk/javascript/pull/3745

File: docs/testing/cypress.mdx [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+  // ...
+});
+```
+
+</Steps>
+
+## Custom Commands
+
+The `@clerk/testing` package also provides [Cypress custom commands](https://docs.cypress.io/api/cypress-api/custom-commands) to sign in/sign out with Clerk in your Cypress tests without having to interact with the UI.
+To use these commands, you must import them in your `cypress/support/commands.ts` file.
```
Comment (alexisintech) [inline docs/testing/cypress.mdx:93]: they _need_ to use them in this specific file? or they _can_ use them like the example shows?

File: docs/testing/cypress.mdx [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+
+### `cy.clerkSignIn`
+
+The `cy.clerkSignIn` command is used to sign in a user using Clerk. This custom command supports only the following first factor strategies:
+
+- Password
+- Phone code
+- Email code
+
+**Note:** Multi-factor authentication is not supported.
```
Comment (alexisintech) [inline docs/testing/cypress.mdx:112]:
```suggestion
Multi-factor authentication is not supported.
```

File: docs/testing/cypress.mdx [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+
+> [!NOTE]
+> This helper internally uses the `setupClerkTestingToken`, so you don't need to call it separately.
```
Comment (alexisintech) [inline docs/testing/cypress.mdx:115]:
```suggestion
> This helper internally uses the `setupClerkTestingToken()` helper, so you don't need to call it separately.
```

File: docs/testing/cypress.mdx [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+
+**Prerequisites**
+
+- You must call `cy.visit` before calling this command.
+- Navigate to a non-protected page that loads Clerk before using this command.
```
Comment (alexisintech) [inline docs/testing/cypress.mdx:120]:
```suggestion
#### Prerequisites

- Before calling this command, you must call `cy.visit`.
- Before using this command, navigate to a non-protected page that loads Clerk.
```

File: docs/testing/cypress.mdx [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+
+**Parameters**
+
+This command accepts an object with the following properties:
+
+- `strategy`: The sign-in strategy. Supported strategies are `'password'`, `'phone_code'`, or `'email_code'`.
+- `identifier`: The user's identifier. This could be a username, a phone number, or an email.
+- `password`: The user's password. This is required only if the strategy is `'password'`.
```
Comment (alexisintech) [inline docs/testing/cypress.mdx:128]:
```suggestion
#### Parameters

This command accepts an object with the following properties:

- `strategy`: The sign-in strategy. Supported strategies are `'password'`, `'phone_code'`, or `'email_code'`.
    - `password`: The command will sign in the user using the provided password and identifier.
    - `phone_code`: You must have a user with a test phone number as an identifier (e.g., `+15555550100`).
    - `email_code`: You must have a user with a test email as an identifier (e.g., `your_email+clerk_test@example.com`).
- `identifier`: The user's identifier. This could be a username, a phone number, or an email.
- `password`: The user's password. This is required only if the `strategy` is set to `'password'`.
```

File: docs/testing/cypress.mdx [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+
+**Strategy Specifics**
+
+- **Password:** The command will sign in the user using the provided password and identifier.
+- **Phone Code:** You must have a user with a test phone number as an identifier (e.g., `+15555550100`).
+- **Email Code:** You must have a user with a test email as an identifier (e.g., `your_email+clerk_test@example.com`).
```
Comment (alexisintech) [inline docs/testing/cypress.mdx:134]: Removed because it's moved to suggestion above

```suggestion
```

File: docs/testing/cypress.mdx [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+
+**Example**
+
```
Comment (alexisintech) [inline docs/testing/cypress.mdx:137]:
```suggestion
#### Example

The following example demonstrates how to use the `cy.clerkSignIn()` command in a test to sign in a user.

```

File: docs/testing/cypress.mdx [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+```
+
+### `cy.clerkSignOut`
+
+The `cy.clerkSignOut` command is used to sign out the current user using Clerk.
+
+**Prerequisites**
+
+- You must call `cy.visit` before calling this command.
+- Navigate to a page that loads Clerk before using this command.
```
Comment (alexisintech) [inline docs/testing/cypress.mdx:154]:
```suggestion
#### Prerequisites

- Before calling this command, yu must call `cy.visit`.
- Before using this command, navigate to a page that loads Clerk.
```

File: docs/testing/cypress.mdx [hunk 18]
```diff
@@ -0,0 +1,191 @@
...
+
+**Parameters**
+
+- `signOutOptions`: A `SignOutOptions` object.
```
Comment (alexisintech) [inline docs/testing/cypress.mdx:158]: Where is the object?? Is there supposed to be a reference link here?

```suggestion
#### Parameters

- `signOutOptions`: A `SignOutOptions` object.
```

File: docs/testing/cypress.mdx [hunk 19]
```diff
@@ -0,0 +1,191 @@
...
+
+**Example**
+
```
Comment (alexisintech) [inline docs/testing/cypress.mdx:161]:
```suggestion
#### Example

The following example demonstrates how to use the `cy.clerkSignOut()` command in a test to sign a user out.
```

File: docs/testing/cypress.mdx [hunk 20]
```diff
@@ -0,0 +1,191 @@
...
+The `cy.clerkLoaded` command asserts that Clerk has been loaded.
+
+**Prerequisites**
+
+- You must call `cy.visit` before calling this command.
+- Navigate to a page that loads Clerk before using this command.
+
+**Example**
+
```
Comment (alexisintech) [inline docs/testing/cypress.mdx:182]:
```suggestion
#### Prerequisites

- Before calling this command, you must call `cy.visit` before calling this command.
- Before using this command, navigate to a page that loads Clerk.

#### Example

The following example demonstrates how to use the `cy.clerkLoaded()` command in a test to assert that Clerk has been loaded.
```

File: docs/testing/cypress.mdx (hunk 10 above)
Comment (anagstef) [reply docs/testing/cypress.mdx:93]: Hmm, they need to use them in their support file, so they'll be loaded before any test is run. 
The support file path for E2E tests could be one of the following:
- cypress/support/e2e.js
- cypress/support/e2e.jsx
- cypress/support/e2e.ts
- cypress/support/e2e.tsx

It's a common practice to have a `commands.ts/js` file in the `support` folder and import it inside the `e2e.ts/js`, that's why I mentioned this file here, but we could rephrase it to be clearer. Do you have any suggestions on this? 🤔 
support file ref: https://docs.cypress.io/guides/core-concepts/writing-and-organizing-tests#Support-file

File: docs/testing/cypress.mdx [hunk 21]
```diff
@@ -0,0 +1,193 @@
...
+  // ...
+});
+```
+
+</Steps>
+
+## Custom Commands
+
+The `@clerk/testing` package also provides [Cypress custom commands](https://docs.cypress.io/api/cypress-api/custom-commands) to sign in/sign out with Clerk in your Cypress tests without having to interact with the UI.
+To use these commands, you must import them in your `cypress/support/commands.ts` file.
```
Comment (anagstef) [inline docs/testing/cypress.mdx:93]:
```suggestion
To use these commands, you must import them in your [Cypress E2E Support file](https://docs.cypress.io/guides/core-concepts/writing-and-organizing-tests#Support-file), which by default is the `cypress/support/e2e.{js,jsx,ts,tsx}` file.
```
@alexisintech What do you think about this change?? 🤔

Comment (alexisintech) [reply docs/testing/cypress.mdx:93]: I like it - I think providing references is helpful

**Review by LekoArts — APPROVED** [review]

File: docs/testing/cypress.mdx [hunk 22]
```diff
@@ -0,0 +1,193 @@
+---
+title: Testing with Cypress
+description: Use Cypress to write end-to-end tests with Clerk.
+---
+
+# Testing with Cypress
+
+> [!WARNING]
+> Our `@clerk/testing` package supports only end-to-end testing with Cypress. Unit tests are not supported.
```
Comment (alexisintech) [inline docs/testing/cypress.mdx:9]: Does it not also support E2E testing with playwright? If so, let's write:

```suggestion
> Our `@clerk/testing` package only supports end-to-end testing. Unit tests are not supported.
```

File: docs/testing/cypress.mdx [hunk 23]
```diff
@@ -0,0 +1,193 @@
...
+
+</Steps>
+
+## Custom Commands
+
+The `@clerk/testing` package also provides [Cypress custom commands](https://docs.cypress.io/api/cypress-api/custom-commands) to sign in/sign out with Clerk in your Cypress tests without having to interact with the UI.
+To use these commands, you must import them in your [Cypress E2E Support file](https://docs.cypress.io/guides/core-concepts/writing-and-organizing-tests#Support-file), which by default is the `cypress/support/e2e.{js,jsx,ts,tsx}` file.
+
+Example:
+```ts {{ filename: 'cypress/support/commands.ts' }}
```
Comment (alexisintech) [inline docs/testing/cypress.mdx:96]: File name should match the default

```suggestion
```ts {{ filename: 'cypress/support/e2e.ts' }}
```

File: docs/testing/cypress.mdx [hunk 24]
```diff
@@ -0,0 +1,193 @@
...
+  // ...
+});
+```
+
+</Steps>
+
+## Custom Commands
+
+The `@clerk/testing` package also provides [Cypress custom commands](https://docs.cypress.io/api/cypress-api/custom-commands) to sign in/sign out with Clerk in your Cypress tests without having to interact with the UI.
+To use these commands, you must import them in your [Cypress E2E Support file](https://docs.cypress.io/guides/core-concepts/writing-and-organizing-tests#Support-file), which by default is the `cypress/support/e2e.{js,jsx,ts,tsx}` file.
```
Comment (alexisintech) [inline docs/testing/cypress.mdx:93]:
```suggestion
To use these commands, you must import them in your [Cypress E2E support file](https://docs.cypress.io/guides/core-concepts/writing-and-organizing-tests#Support-file), as shown in the following example.
```

File: docs/testing/cypress.mdx [hunk 25]
```diff
@@ -0,0 +1,193 @@
...
+> This helper internally uses the `setupClerkTestingToken()` helper, so you don't need to call it separately.
+
+#### Prerequisites
+
+- Before calling this command, you must call `cy.visit`.
+- Before using this command, navigate to a non-protected page that loads Clerk.
+
+#### Parameters
+
+This command accepts an object with the following properties:
```
Comment (alexisintech) [inline docs/testing/cypress.mdx:124]:
```suggestion
`cy.clerkSignIn` accepts an object with the following properties:
```

File: docs/testing/cypress.mdx [hunk 26]
```diff
@@ -0,0 +1,193 @@
...
+The `cy.clerkSignOut` command is used to sign out the current user using Clerk.
+
+#### Prerequisites
+
+- Before calling this command, yu must call `cy.visit`.
+- Before using this command, navigate to a page that loads Clerk.
+
+#### Parameters
+
+- `signOutOptions`: A `SignOutOptions` object.
```
Comment (alexisintech) [inline docs/testing/cypress.mdx:158]: I had left a note with a comment before, maybe you missed it - but where is the `SignOutOptions` object? Is there a reference link for it, or do we need to provide it here.

File: docs/testing/cypress.mdx (hunk 22 above)
Comment (anagstef) [reply docs/testing/cypress.mdx:9]: Hmm, actually yes, we only support E2E tests, but Playwright has no unit tests either way. So, I added this note only here because it affects mainly Cypress, but we can make it more generic as you suggest, no problem.

File: docs/testing/cypress.mdx (hunk 26 above)
Comment (anagstef) [reply docs/testing/cypress.mdx]: I missed this one. I added a link to it, also marked it as optional

File: docs/testing/cypress.mdx [hunk 27]
```diff
@@ -0,0 +1,192 @@
...
+The `cy.clerkSignOut` command is used to sign out the current user using Clerk.
+
+#### Prerequisites
+
+- Before calling this command, yu must call `cy.visit`.
+- Before using this command, navigate to a page that loads Clerk.
+
+#### Parameters
+
+- `signOutOptions`: A [`SignOutOptions`](https://clerk.com/docs/components/unstyled/sign-out-button#sign-out-options) object. This parameter is optional.
```
Comment (alexisintech) [inline docs/testing/cypress.mdx:157]:
```suggestion
`cy.clerkSignOut` accepts an optional parameter `signOutOptions`, which includes the following properties:

| Name | Type | Description |
| --- | --- | --- |
| `sessionId?` | `string` | The ID of a specific session to sign out of. Useful for multi-session applications. |
| `redirectUrl?` | `string` | The redirect URL to navigate to after sign out is complete. |
```
